src/compiler/interface2.jl

```diff
@@ -3,7 +3,86 @@ using IRTools.Inner: argnames!, update!
 
 ignore(T) = all(T -> T <: Type, T.parameters)
 
-@generated function _pullback(ctx::AContext, f, args...)
+
+function _pullback(ctx::AContext, f, args...)
+  if chainrules_blacklist(f, args...)
+    # then don't even consider using ChainRules
+    return _pullback_via_source2source(ctx, f, args...)
+  end
+
+  res = ChainRules.rrule(f, args...)
+  if res === nothing
+    # No ChainRule defined, time to do the source tranform
+    return _pullback_via_source2source(ctx, f, args...)
+  else
+    # Can just use ChainRule answer
+    y, pb = res
+    return y, _pullback_via_chainrules(pb)
+  end
+end
+
+#=="""
+  chainrules_blacklist(f, args...,)
+
+This is used to disable the use of ChainRule's definitions
+for particular functions/methods.
+
+It is not required if a Zygote rule has already been defined directly.
+"""==#
+chainrules_blacklist(f, args...) = false
+
+# ChainRules does higher-order functions badly
+# see https://github.com/JuliaDiff/ChainRules.jl/issues/122
+chainrules_blacklist(::typeof(map), args...) = true
+chainrules_blacklist(::typeof(broadcast), args...) = true
+chainrules_blacklist(::typeof(mapreduce), args...) = true
+chainrules_blacklist(::typeof(mapfoldl), args...) = true
+chainrules_blacklist(::typeof(mapfoldr), args...) = true
+chainrules_blacklist(::typeof(sum), f, x::AbstractArray{<:Real}) = true
+# Except for sum(abs2, xs), that is fine
+chainrules_blacklist(::typeof(sum), ::typeof(abs2), x::AbstractArray{<:Real}) = false
+
+# ChainRules current Wirtinger deriviative is not compatible
+# reconsider after https://github.com/JuliaDiff/ChainRulesCore.jl/pull/29
+chainrules_blacklist(::typeof(abs), ::Complex) = true
```

Comment:
Use metaprogramming?

Reply:
Could do.
But I feel like wouldn't be much shorter and would be less clear

Reply:
Maybe should do.
This also could be done with `Union`.

`uniontypeof(fs...) = Union{typeof.(fs)...}`